app/src/main/java/org/mozilla/fenix/tabstray/browser/TabSorter.kt

```diff
@@ -0,0 +1,144 @@
+    override fun onTabsChanged(position: Int, count: Int) = Unit
+    override fun onTabsInserted(position: Int, count: Int) = Unit
+    override fun onTabsMoved(fromPosition: Int, toPosition: Int) = Unit
+    override fun onTabsRemoved(position: Int, count: Int) = Unit
+}
+
+private fun List<Tab>.findSelectedIndex(tabId: String?): Int {
+    val id = tabId ?: return -1
+    return indexOfFirst { it.id == id }
+}
+
+/**
+ * Returns a list of inactive tabs based on our preferences.
+ */
+private fun List<Tab>.getInactiveTabs(context: Context): List<Tab> {
+    val inactiveTabsEnabled = context.settings().inactiveTabsAreEnabled
+    return if (inactiveTabsEnabled) {
+        filter { !it.isActive(maxActiveTime) }
+    } else {
+        emptyList()
+    }
+}
+
+/**
+ * Returns a list of search term tabs based on our preferences.
+ */
+private fun List<Tab>.getSearchGroupTabs(context: Context): List<Tab> {
+    val inactiveTabsEnabled = context.settings().inactiveTabsAreEnabled
+    val tabGroupsEnabled = FeatureFlags.tabGroupFeature
+    return when {
+        tabGroupsEnabled && inactiveTabsEnabled ->
+            filter { it.searchTerm.isNotBlank() && it.isActive(maxActiveTime) }
+
+        tabGroupsEnabled ->
+            filter { it.searchTerm.isNotBlank() }
+
+        else -> emptyList()
+    }
+}
+
+/**
+ * Returns true if a tab has not been selected since [maxActiveTime].
+ *
+ * N.B: This is duplicated from [TabSessionState.isActive(Long)] to work for [Tab].
+ *
+ * See also: https://github.com/mozilla-mobile/android-components/issues/11012
+ */
+private fun Tab.isActive(maxActiveTime: Long): Boolean {
+    val lastActiveTime = maxOf(lastAccess, createdAt)
+    val now = System.currentTimeMillis()
+    return (now - lastActiveTime <= maxActiveTime)
+}
+
+/**
+ * Creates a list of grouped search term tabs sorted by last access time and a list of tabs
+ * that have search terms but would only create groups with a single tab.
+ *
+ * N.B: This is duplicated from [List<TabSessionState>.toSearchGroup()] to work for [Tab].
+ *
+ * See also: https://github.com/mozilla-mobile/android-components/issues/11012
+ */
+private fun List<Tab>.toSearchGroups(): Pair<List<TabGroupAdapter.Group>, List<Tab>> {
+    val data = groupBy { it.searchTerm.lowercase() }
+
+    val groupings = data.map { mapEntry ->
+        // Uppercase since we use it for the title.
+        val searchTerm = mapEntry.key.replaceFirstChar(Char::uppercase)
+        val groupTabs = mapEntry.value
+
+        // Calculate when the group was last used.
+        val groupMax = groupTabs.fold(0L) { acc, tab ->
+            max(tab.lastAccess, acc)
+        }
+
+        TabGroupAdapter.Group(
+            title = searchTerm,
+            tabs = groupTabs,
+            lastAccess = groupMax
+        )
+    }
+
+    val groups = groupings.filter { it.tabs.size > 1 }.sortedBy { it.lastAccess }
+    val remainderTabs = (groupings - groups).flatMap { it.tabs }
```

Comment:
doing math operations on collections is always a neat trick

Reply:
+1 to Roger for this.

It simplifies our logic too from using separate filters for each feature.